apollo-ios/Sources/Apollo/ExecutionSources/CacheDataExecutionSource.swift

```diff
@@ -67,6 +68,60 @@ struct CacheDataExecutionSource: GraphQLExecutionSource {
       }
     }
   }
+  
+  private func resolveCacheKey(
+    with info: FieldExecutionInfo,
+    on object: Record
+  ) throws -> AnyHashable? {
+    
+    let fieldTypename = typename(for: info.field)
+    
+    // Programmatic field policy checks
+    switch info.field.type {
+    case .nonNull(let innerType):
+      if case .list(_) = innerType.namedType {
+        if let cacheKeys = info.parentInfo.schema.configuration.cacheKeys(for: info.field, variables: info.parentInfo.variables, path: info.responsePath) {
+          return cacheKeys.map { "\($0.uniqueKeyGroup ?? fieldTypename):\($0.id)" }.map { object[$0] }
+        }
+      }
+      else {
+        if let cacheKey = info.parentInfo.schema.configuration.cacheKey(for: info.field, variables: info.parentInfo.variables, path: info.responsePath) {
+          return object["\(cacheKey.uniqueKeyGroup ?? fieldTypename):\(cacheKey.id)"]
+        }
+      }
+    case .list(_):
+      if let cacheKeys = info.parentInfo.schema.configuration.cacheKeys(for: info.field, variables: info.parentInfo.variables, path: info.responsePath) {
+        return cacheKeys.map { "\($0.uniqueKeyGroup ?? fieldTypename):\($0.id)" }.map { object[$0] }
+      }
+    default:
+      if let cacheKey = info.parentInfo.schema.configuration.cacheKey(for: info.field, variables: info.parentInfo.variables, path: info.responsePath) {
+        return object["\(cacheKey.uniqueKeyGroup ?? fieldTypename):\(cacheKey.id)"]
+      }
+    }
+    
+    // Directive based field policy checks
+    if let fieldPolicyResult = FieldPolicyEvaluator(field: info.field, variables: info.parentInfo.variables).resolveFieldPolicy() {
+      switch fieldPolicyResult {
+      case .single(let key):
+        return object["\(key.uniqueKeyGroup ?? fieldTypename):\(key.id)"]
+      case .list(let keys):
+        return keys.map { object["\($0.uniqueKeyGroup ?? fieldTypename):\($0.id)"] }
+      }
+    }
+    
+    let key = try info.cacheKeyForField()
+    return object[key]
+  }
+  
+  private func typename(for field: Selection.Field) -> String {
+    switch field.type.namedType {
+    case .object(let selectionSetType):
+      return selectionSetType.__parentType.__typename
+    default:
+      break
```

Comment:
Just `return ""` in the default case instead of breaking.

Reply:
Actually, if the field isn't an object field, we have a serious coding error here, don't we? We shouldn't be continuing on with an empty typename added to the cache key. I think this should be an `assertionFailure` so we catch it early.